### ISSUE TITLE
e2e flake: deployment should support rollback when there's replica set with no revision [wrong event]

### ISSUE THREAD

**Issue by kargakis** [kubernetes/kubernetes#21467]
```
â€¢ Failure [54.571 seconds]
Deployment
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:68
  deployment should support rollback when there's replica set with no revision [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:67

  Expected error:
      <*errors.StatusError | 0xc20844e300>: {
          ErrStatus: {
              TypeMeta: {Kind: "Status", APIVersion: "v1"},
              ListMeta: {SelfLink: "", ResourceVersion: ""},
              Status: "Failure",
              Message: "deployments.extensions \"nginx-deployment\" cannot be updated: the object has been modified; please apply your changes to the latest version and try again",
              Reason: "Conflict",
              Details: {
                  Name: "nginx-deployment",
                  Group: "extensions",
                  Kind: "deployments",
                  Causes: nil,
                  RetryAfterSeconds: 0,
              },
              Code: 409,
          },
      }
      deployments.extensions "nginx-deployment" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
  not to have occurred

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:168
```

**Comment by kargakis** [comment]
cc: @janetkuo @mqliang @bgrant0607

**Comment by bgrant0607** [comment]
@kargakis I think this is the same problem fixed by #21346

**Comment by brendandburns** [comment]
@bgrant0607 this is different, since that PR merged and this just flaked again.

**Comment by bgrant0607** [comment]
```
	err = reaper.Stop(ns, deployment.Name, timeout, api.NewDeleteOptions(0))
```

Could be the same problem as #21544

**Comment by janetkuo** [comment]
There's another different failure:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:67
Expected error:
    <*errors.errorString | 0xc208539740>: {
        s: "old replica sets are not fully scaled down",
    }
    old replica sets are not fully scaled down
not to have occurred
```

**Comment by bgrant0607** [comment]
@janetkuo That failure mode was reported on #21491

**Comment by bgrant0607** [comment]
@janetkuo No, sorry, that's a different one.

**Comment by bgrant0607** [comment]
#21463 is the other one with "old replica sets are not fully scaled down"

**Comment by janetkuo** [comment]
This failure is also related to DeploymentReaper.Stop.

**Comment by janetkuo** [comment]
#21572 should fix this

**Comment by bgrant0607** [comment]
Failure with a new symptom:
```
Deployment
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:69
  deployment should support rollback when there's replica set with no revision [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:68

  Expected
      <string>: ScalingReplicaSet
  to equal
      <string>: DeploymentRollback

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:761
```
Buckets/kubernetes-jenkins/logs/kubernetes-e2e-gce/11953

First of all, we shouldn't be relying on events in e2e tests #21511. But we certainly shouldn't be relying on specific events occurring in a specific order.
```
Expect(events.Items[4].Reason).Should(Equal(deploymentutil.RollbackDone))
```

```
Feb 20 11:17:11.786: INFO: event for nginx-deployment: {deployment-controller } ScalingReplicaSet: Scaled up replica set nginx-deployment-rjmhn to 1
Feb 20 11:17:11.786: INFO: event for nginx-deployment: {deployment-controller } DeploymentRollbackRevisionNotFound: Unable to find last revision.
Feb 20 11:17:11.786: INFO: event for nginx-deployment: {deployment-controller } ScalingReplicaSet: Scaled up replica set nginx-deployment-3dp8s to 1
Feb 20 11:17:11.786: INFO: event for nginx-deployment: {deployment-controller } ScalingReplicaSet: Scaled down replica set nginx-deployment-rjmhn to 0
Feb 20 11:17:11.786: INFO: event for nginx-deployment: {deployment-controller } ScalingReplicaSet: Scaled up replica set nginx-deployment-6ige4 to 1
Feb 20 11:17:11.786: INFO: event for nginx-deployment: {deployment-controller } DeploymentRollback: Rolled back deployment "nginx-deployment" to revision 1
Feb 20 11:17:11.786: INFO: event for nginx-deployment: {deployment-controller } ScalingReplicaSet: Scaled down replica set nginx-deployment-3dp8s to 0
Feb 20 11:17:11.786: INFO: event for nginx-deployment: {deployment-controller } ScalingReplicaSet: Scaled down replica set nginx-deployment-6ige4 to 0
```

**Comment by bgrant0607** [comment]
We should report DeploymentRollbackRevisionNotFound in deployment status somehow, also.

**Comment by bgrant0607** [comment]
For both DeploymentRollbackRevisionNotFound and DeploymentRollback, we could wait until rollbackTo is cleared, and then check the Deployment's pod template.

**Comment by AdoHe** [comment]
Another occurrence:
```
Deployment
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:69
  deployment should support rollback when there's replica set with no revision [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:68

  Expected error:
      <*errors.errorString | 0xc2084cb370>: {
          s: "old replica sets are not fully scaled down",
      }
      old replica sets are not fully scaled down
  not to have occurred

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:755
```
[https://console.cloud.google.com/m/cloudstorage/b/kubernetes-jenkins/o/pr-logs/pull/21369/kubernetes-pull-build-test-e2e-gce/29640/build-log.txt](https://console.cloud.google.com/m/cloudstorage/b/kubernetes-jenkins/o/pr-logs/pull/21369/kubernetes-pull-build-test-e2e-gce/29640/build-log.txt)

**Comment by janetkuo** [comment]
Filed #21710 to fix the "wrong event" failure mode.

**Comment by janetkuo** [comment]
#21715 should fix the "old replica sets are not fully scaled down" failure mode.

**Comment by bgrant0607** [comment]
Merging so we can watch the failure rate over night

**Comment by bgrant0607** [comment]
Both of those fixes are now merged

**Comment by bgrant0607** [comment]
This most recently failed with:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:68
Expected
    <string>: ScalingReplicaSet
to equal
    <string>: DeploymentRollbackRevisionNotFound
```
which should be fixed by #21710, which was merged after #21715.

**Comment by janetkuo** [comment]
This should be fixed.
